File: ex_2_7_2.py
import pandas as pd;
import skfuzzy as fuzzy;
import numpy as np;
from skfuzzy import control as ctrl;
import matplotlib.pyplot as plt;

#EXERCÍCIO 2.7.2

#---------------------------------------------------------
# Definindo os domínios

massa = ctrl.Antecedent(np.arange(47, 81.1, 0.1), 'Massa');
altura = ctrl.Antecedent(np.arange(157, 184, 1), 'Altura');
grau_risco = ctrl.Consequent(np.arange(18, 36, 1), 'Grau de Risco');

#---------------------------------------------------------
# Definindo as faixas de pertinência

massa['Baixa'] = fuzzy.trapmf(massa.universe, [47, 47, 55, 64]);
massa['Média Baixa'] = fuzzy.trimf(massa.universe, [50, 59, 68]);
massa['Média'] = fuzzy.trimf(massa.universe, [53, 62.5, 72]);
massa['Média Alta'] = fuzzy.trimf(massa.universe, [56, 66.5, 77]);
massa['Alta'] = fuzzy.trapmf(massa.universe, [59, 70, 81, 81]);

altura['Baixa'] = fuzzy.trapmf(altura.universe, [157, 157, 160, 163]);
altura['Média Baixa'] = fuzzy.trimf(altura.universe, [162, 165, 168]);
altura['Média'] = fuzzy.trimf(altura.universe, [167, 170, 173]);
altura['Média Alta'] = fuzzy.trimf(altura.universe, [172, 175, 178]);
altura['Alta'] = fuzzy.trapmf(altura.universe, [177, 180, 183, 183]);

grau_risco['Saudável'] = fuzzy.trapmf(grau_risco.universe, [18, 18, 23, 25]);
grau_risco['Moderado'] = fuzzy.trapmf(grau_risco.universe, [24, 25, 27, 30]);
grau_risco['Alto'] = fuzzy.trapmf(grau_risco.universe, [29, 30, 35, 35]);

#---------------------------------------------------------
# Salvando os gráficos em png
'''
massa.view();
altura.view();
grau_risco.view();
'''
#---------------------------------------------------------
# Definindo as regras

conjunto_regras = [];

conjunto_regras.append(ctrl.Rule(massa['Baixa'] & altura['Média'], grau_risco['Saudável']));
conjunto_regras.append(ctrl.Rule(massa['Baixa'] & altura['Média Baixa'], grau_risco['Saudável']));
conjunto_regras.append(ctrl.Rule(massa['Baixa'] & altura['Média'], grau_risco['Saudável']));
conjunto_regras.append(ctrl.Rule(massa['Baixa'] & altura['Média Alta'], grau_risco['Saudável']));
conjunto_regras.append(ctrl.Rule(massa['Baixa'] & altura['Alta'], grau_risco['Saudável']));

conjunto_regras.append(ctrl.Rule(massa['Média Baixa'] & altura['Média'], grau_risco['Moderado']));
conjunto_regras.append(ctrl.Rule(massa['Média Baixa'] & altura['Média Baixa'], grau_risco['Saudável']));
conjunto_regras.append(ctrl.Rule(massa['Média Baixa'] & altura['Média'], grau_risco['Saudável']));
conjunto_regras.append(ctrl.Rule(massa['Média Baixa'] & altura['Média Alta'], grau_risco['Saudável']));
conjunto_regras.append(ctrl.Rule(massa['Média Baixa'] & altura['Alta'], grau_risco['Saudável']));

conjunto_regras.append(ctrl.Rule(massa['Média'] & altura['Média'], grau_risco['Moderado']));
conjunto_regras.append(ctrl.Rule(massa['Média'] & altura['Média Baixa'], grau_risco['Moderado']));
conjunto_regras.append(ctrl.Rule(massa['Média'] & altura['Média'], grau_risco['Saudável']));
conjunto_regras.append(ctrl.Rule(massa['Média'] & altura['Média Alta'], grau_risco['Saudável']));
conjunto_regras.append(ctrl.Rule(massa['Média'] & altura['Alta'], grau_risco['Saudável']));

conjunto_regras.append(ctrl.Rule(massa['Média Alta'] & altura['Média'], grau_risco['Moderado']));
conjunto_regras.append(ctrl.Rule(massa['Média Alta'] & altura['Média Baixa'], grau_risco['Moderado']));
conjunto_regras.append( ctrl.Rule(massa['Média Alta'] & altura['Média'], grau_risco['Moderado']));
conjunto_regras.append(ctrl.Rule(massa['Média Alta'] & altura['Média Alta'], grau_risco['Saudável']));
conjunto_regras.append(ctrl.Rule(massa['Média Alta'] & altura['Alta'], grau_risco['Saudável']));

conjunto_regras.append(ctrl.Rule(massa['Alta'] & altura['Média'], grau_risco['Alto']));
conjunto_regras.append(ctrl.Rule(massa['Alta'] & altura['Média Baixa'], grau_risco['Moderado']));
conjunto_regras.append(ctrl.Rule(massa['Alta'] & altura['Média'], grau_risco['Moderado']));
conjunto_regras.append(ctrl.Rule(massa['Alta'] & altura['Média Alta'], grau_risco['Moderado']));
conjunto_regras.append(ctrl.Rule(massa['Alta'] & altura['Alta'], grau_risco['Saudável']));

#---------------------------------------------------------
# Criando o sistema de regras do controle

grau_risco_ctrl = ctrl.ControlSystem(conjunto_regras);
grau_risco_simulacao = ctrl.ControlSystemSimulation(grau_risco_ctrl);

#---------------------------------------------------------
# Inputs

grau_risco_simulacao.input['Massa'] = 59.0;  # (float(input()))
grau_risco_simulacao.input['Altura'] = 164; # (int(input()))

#---------------------------------------------------------
# Saída

grau_risco_simulacao.compute();
print(grau_risco_simulacao.output['Grau de Risco']);
# ERRO de ~0.5

File: ex272.m
% EXERCÍCIO 2.7.2
fis_272 = mamfis(Name="Grau de Risco de Obesidade");

% Definindo os domínios
fis_272 = addInput(fis_272,[47 81],Name="Massa");
fis_272 = addInput(fis_272,[157 183],Name="Altura");
fis_272 = addOutput(fis_272,[18 35],Name="Grau");

% Definindo as faixas de pertinência
fis_272 = addMF(fis_272,"Massa","trapmf",[47 47 55.5 64],Name="Baixa");
fis_272 = addMF(fis_272,"Massa","trimf",[50 59 68],Name="Média Baixa");
fis_272 = addMF(fis_272,"Massa","trimf",[53 62.5 72],Name="Média");
fis_272 = addMF(fis_272,"Massa","trimf",[56 66.5 77],Name="Média Alta");
fis_272 = addMF(fis_272,"Massa","trapmf",[59 70 81 81],Name="Alta");

fis_272 = addMF(fis_272,"Altura","trapmf",[157 157 160 163],Name="Baixa");
fis_272 = addMF(fis_272,"Altura","trimf",[162 165 168],Name="Média Baixa");
fis_272 = addMF(fis_272,"Altura","trimf",[167 170 173],Name="Média");
fis_272 = addMF(fis_272,"Altura","trimf",[172 175 178],Name="Média Alta");
fis_272 = addMF(fis_272,"Altura","trapmf",[177 180 183 183],Name="Alta");

fis_272 = addMF(fis_272,"Grau","trapmf",[18 18 22.5 25],Name="Saudável");
fis_272 = addMF(fis_272,"Grau","trapmf",[24 25.5 28 30],Name="Moderado");
fis_272 = addMF(fis_272,"Grau","trapmf",[29 30.5 35 35],Name="Alto");

% Definindo as regras [massa altura grau peso operação]
ruleList_272 = [
    1 1 1 1 1  % Massa Baixa, Altura Baixa -> Saudável
    1 2 2 1 1  % Massa Baixa, Altura Média Baixa -> Moderado
    1 3 2 1 1  % Massa Baixa, Altura Média -> Moderado
    1 4 2 1 1  % Massa Baixa, Altura Média Alta -> Moderado
    1 5 3 1 1  % Massa Baixa, Altura Alta -> Alto
    
    2 1 1 1 1  % Massa Média Baixa, Altura Baixa -> Saudável
    2 2 1 1 1  % Massa Média Baixa, Altura Média Baixa -> Saudável
    2 3 2 1 1  % Massa Média Baixa, Altura Média -> Moderado
    2 4 2 1 1  % Massa Média Baixa, Altura Média Alta -> Moderado
    2 5 2 1 1  % Massa Média Baixa, Altura Alta -> Moderado
    
    3 1 1 1 1  % Massa Média, Altura Baixa -> Saudável
    3 2 1 1 1  % Massa Média, Altura Média Baixa -> Saudável
    3 3 1 1 1  % Massa Média, Altura Média -> Saudável
    3 4 2 1 1  % Massa Média, Altura Média Alta -> Moderado
    3 5 2 1 1  % Massa Média, Altura Alta -> Moderado
    
    4 1 1 1 1  % Massa Média Alta, Altura Baixa -> Saudável
    4 2 1 1 1  % Massa Média Alta, Altura Média Baixa -> Saudável
    4 3 1 1 1  % Massa Média Alta, Altura Média -> Saudável
    4 4 1 1 1  % Massa Média Alta, Altura Média Alta -> Saudável
    4 5 2 1 1  % Massa Média Alta, Altura Alta -> Moderado
    
    5 1 1 1 1  % Massa Alta, Altura Baixa -> Saudável
    5 2 1 1 1  % Massa Alta, Altura Média Baixa -> Saudável
    5 3 1 1 1  % Massa Alta, Altura Média -> Saudável
    5 4 1 1 1  % Massa Alta, Altura Média Alta -> Saudável
    5 5 1 1 1  % Massa Alta, Altura Alta -> Saudável
];

fis_272 = addRule(fis_272, ruleList_272);

% Inputs
massa = 59;
altura = 164;

% Saída
grau = evalfis(fis_272, [massa altura]);